distributor/cmd/internal/http/server.go

```diff
@@ -0,0 +1,136 @@
+	"google.golang.org/grpc/status"
+)
+
+// Server is the core handler implementation of the witness.
+type Server struct {
+	d *distributor.Distributor
+}
+
+// NewServer creates a new server.
+func NewServer(d *distributor.Distributor) *Server {
+	return &Server{
+		d: d,
+	}
+}
+
+// update handles requests to update checkpoints.
+func (s *Server) update(w http.ResponseWriter, r *http.Request) {
+	v := mux.Vars(r)
+	logID := v["logid"]
+	witID := v["witid"]
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("cannot read request body: %v", err.Error()), http.StatusBadRequest)
+		return
+	}
+	if err := s.d.Distribute(r.Context(), logID, witID, body); err != nil {
+		glog.Warningf("failed to update to new checkpoint: %v", err)
+		http.Error(w, "failed to update to new checkpoint", httpForCode(status.Code(err)))
+		return
+	}
+}
+
+// getCheckpointN returns a checkpoint stored for a given log with the specified number of witnesses.
+func (s *Server) getCheckpointN(w http.ResponseWriter, r *http.Request) {
+	v := mux.Vars(r)
+	logID := v["logid"]
+	numSigs, err := strconv.ParseUint(v["numsigs"], 10, 32)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to parse number of signatures: %v", err), http.StatusBadRequest)
+		return
+	}
+	// Get the signed checkpoint from the witness.
+	chkpt, err := s.d.GetCheckpointN(r.Context(), logID, uint32(numSigs))
+	if err != nil {
+		glog.Warningf("failed to get checkpoint: %v", err)
+		http.Error(w, "failed to get checkpoint", httpForCode(status.Code(err)))
+		return
+	}
+	w.Header().Set("Content-Type", "text/plain")
+	w.Write(chkpt)
+}
+
+// getCheckpointWitness returns the latest checkpoint stored for a given log by the given witness.
+func (s *Server) getCheckpointWitness(w http.ResponseWriter, r *http.Request) {
+	v := mux.Vars(r)
+	logID := v["logid"]
+	witID := v["witid"]
+
+	// Get the signed checkpoint from the witness.
+	chkpt, err := s.d.GetCheckpointWitness(r.Context(), logID, witID)
+	if err != nil {
+		glog.Warningf("failed to get checkpoint: %v", err)
+		http.Error(w, "failed to get checkpoint", httpForCode(status.Code(err)))
+		return
+	}
+	w.Header().Set("Content-Type", "text/plain")
+	w.Write(chkpt)
+}
+
+// getLogs returns a list of all logs the witness is aware of.
+func (s *Server) getLogs(w http.ResponseWriter, r *http.Request) {
+	logs, err := s.d.GetLogs(r.Context())
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get log list: %v", err), http.StatusInternalServerError)
+		return
+	}
+	logList, err := json.Marshal(logs)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to convert log list to JSON: %v", err), http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", "text/json")
+	w.Write(logList)
+}
+
+// RegisterHandlers registers HTTP handlers for witness endpoints.
+func (s *Server) RegisterHandlers(r *mux.Router) {
+	logStr := "{logid:[a-zA-Z0-9-]+}"
+	witStr := "{witid:[a-zA-Z0-9-.]+}"
```

Comment:
This looks like it'll allow any character one or more times, is that intended?

Reply:
Good call, I've escaped the dot for now. This is a bit hacky really. Do we have a better dictionary for what might be legal characters in a vkey name?

Reply:
https://cs.opensource.google/go/x/mod/+/refs/tags/v0.10.0:sumdb/note/note.go;drc=7c05a442b7c1d1a107879b4a090bb5a38d3774a1;l=238